rust/src/io/exec/scan.rs

```diff
@@ -25,20 +25,57 @@ use datafusion::physical_plan::{
     ExecutionPlan, Partitioning, RecordBatchStream, SendableRecordBatchStream,
 };
 use futures::stream::Stream;
-use tokio::sync::mpsc::{self, Receiver};
-use tokio::task::JoinHandle;
+use futures::{stream, Future};
+use futures::{StreamExt, TryStreamExt};
 
-use crate::dataset::fragment::FileFragment;
+use crate::dataset::fragment::{FileFragment, FragmentReader};
 use crate::dataset::{Dataset, ROW_ID};
 use crate::datatypes::Schema;
 use crate::format::Fragment;
 
+async fn open_file(
+    file_fragment: FileFragment,
+    projection: Arc<Schema>,
+    with_row_id: bool,
+) -> Result<FragmentReader> {
+    let mut reader = file_fragment.open(projection.as_ref()).await?;
+    if with_row_id {
+        reader.with_row_id();
+    };
+    Ok(reader)
+}
+
+/// Convert a [`FragmentReader`] into a [`Stream`] of [`RecordBatch`].
+fn scan_batches(
+    reader: FragmentReader,
+    read_size: usize,
+) -> Pin<Box<dyn Stream<Item = impl Future<Output = Result<RecordBatch>>> + Send + 'static>> {
+    // To make sure the reader lives long enough, we put it in an Arc.
+    let reader = Arc::new(reader);
+    let reader2 = reader.clone();
+
+    let read_params_iter = (0..reader.num_batches()).flat_map(move |batch_id| {
+        let rows_in_batch = reader.num_rows_in_batch(batch_id);
+        (0..rows_in_batch)
+            .step_by(read_size)
+            .map(move |start| (batch_id, start..min(start + read_size, rows_in_batch)))
+    });
+    let batch_stream = stream::iter(read_params_iter).map(move |(batch_id, range)| {
+        let reader = reader2.clone();
+        async move {
+            reader
+                .read_batch(batch_id, range)
+                .await
+                .map_err(|e| DataFusionError::from(e))
+        }
+    });
```

Comment:
Do you need a `buffered/unbuffered` here to control the number of parallelism?

Reply:
That's configured in the body of `try_new()`, at where your other comment is. This method just defines the stream of futures.